core/src/transform/desugar_destructuring.rs

```diff
@@ -47,13 +48,54 @@ pub fn desugar_fun(mut pat: Pattern, body: RichTerm) -> Term {
 
 /// Desugar a destructuring let-binding.
 ///
-/// A let-binding `let <pat> = bound in body` is desugared to `<bound> |> match { <pat> => body }`.
-pub fn desugar_let(pattern: Pattern, bound: RichTerm, body: RichTerm) -> Term {
+/// A let-binding `let <pat1> = <bound1>, <pat2> = <bound2> in body` is desugared to
+/// `[<bound1>, <bound2>] |> match { [<pat1>, <pat2>] => body }`. If there is only one
+/// pattern, we drop the arrays in the pattern and the bound (so, `<bound1> |> match { <pat1> => body }`).
+pub fn desugar_let(
+    bindings: impl IntoIterator<Item = (Pattern, RichTerm)>,
+    body: RichTerm,
+) -> Term {
     // the position of the match expression is used during error reporting, so we try to provide a
     // sensible one.
-    let match_expr_pos = pattern.pos.fuse(bound.pos);
+    let (bounds_pos, patterns, patterns_pos, bound) = bindings.into_iter().fold(
+        (TermPos::None, Vec::new(), TermPos::None, Vec::new()),
+        |(mut bounds_pos, mut patterns, mut patterns_pos, mut bound), (pat, rt)| {
+            bounds_pos = bounds_pos.fuse(rt.pos);
+            patterns_pos = patterns_pos.fuse(pat.pos);
+
+            patterns.push(pat);
+            bound.push(rt);
+
+            (bounds_pos, patterns, patterns_pos, bound)
+        },
+    );
```

Comment:
It's just a nitpick/suggestion so feel free to ignore, but as much as a I'm a functional programmer at heart, I found folds to be a bit artificial when we're just mutating the accumulator to return it unchanged (unchanged in the sense of it's the same pointer). I find it less readable, especially with 4 arguments, because you have the somehow the drawbacks of the imperative and functional world: you have to pass a lot of stuff explicitly along iterations (taking as argument of the folded function and returning it), and you're mutating it anyway (although here you could argue that `bound_pos` and `pattern_mos` aren't mutated but immutably updated). All in all, I think I would use a good old `for` in this case.